### DISCUSSION TITLE
Variable input/output relationships for units

### DISCUSSION THREAD

**Discussion by LuckyLuke567** [spine-tools/SpineOpt.jl#454] [Ideas]
Hi Spine community,

I was wondering if it is possible in Spine to allow a unit to use (and produce) different combinations of fuels, i.e.:

- a combined-heat-and-power plant which is able to produce heat and electricity, but also electricity alone, or
-  a fuel cell which can be fed with different blends of natural gas and hydrogen (30-70, 40-60, 50-50...). 

I wanted also to ask if these option, which is something like the MODE_OF_OPERATION in OSeMOSYS, can be subject of optimisation.

**Comment by DillonJ** [comment]
Hi there!

Thanks for your interest in Spine! In SpineOpt a unit can have an arbitrary number of inputs and an arbitrary number of outputs with piecewise linear relationships defined between any and all inputs/outputs - so it is very flexible.

Regarding the MODE_OF_OPERATION question - I know of OSeMOSYS but I'm not very familliar with its implementation. I imagine this is some binary or integer variable that decides what outputs etc can be produced.

In SpineOpt, the `units_online` variable which can be continuous, binary or integer, determines if the unit is online or not. If online, it can operate with the linear constraints dictating the relationships between inputs and outputs. So both of your examples above are possible.

However in SpineOpt, the entire unit is either on or off. If you need to implement discrete modes of operation, it may be necessary to define a number of sub units - each of which represents a specific mode of operation. Then you can create a user constraint to allow only one of the units to operate at any one time.

Please let me know if you'd like any further information

**Reply by LuckyLuke567** [reply]
Dear DillonJ, thank you for your fast and insightful response.

I wil try to implement your suggestions in my model, and see what happens. I will post the outcome here, if needed. 

I see only one potential drawback. Taking as example a long-term investment problem, with a scenario which foresees a decline in the cost of hydrogen, if I was to define several units for the abovementioned fuel cell it would be hard to observe a switch from low to hydrogen-rich blends through the years, as changing the blend would require investing in a new unit, whereas in reality the same fuel cell can operate with different blends.

Nevertheless, I will try it out and let you know,
Thanks again.

P.S. I am not 100% sure about the implementation of `MODE_OF_OPERATION` in OSeMOSYS, but it remains in the range of linear programming. It's not even a variable, nor a parameter: it is called a set in jargon, but I would not know how to translate it in another framework. It allows you to define any number of modes of operation, that is: the same technology may have different input/output relationships, operating costs etc. for each mode. The proportion of utilisation of each mode is then optimised by the solver.

**Comment by DillonJ** [comment]
You could model a unit with two inputs, hydrogen and natural gas, and one output, electricity. You use the fix_ratio_unit_flow_out_in for each.

You could then say `candidate_units` = 10 and `unit_capacity` = [your total capacity]/10

This would force the model to invest in the Hydrogen and Natural Gas Elements in 10% capacity blocks. I guess you could also add a constraint so that the total units_invested = 10 - thus making the ratio the investment variable.

There could be many many ways to do this, but I'm sure it's possible!

**Reply by DillonJ** [reply]
Actually - this could work better: You could make the unit_capacity = 1 for the hydrogen and natural gas components and make the investment variable continuous. Then the model can simultaneously decide how much capacity to build and also the ratio between hydrogen and natural gas

**Comment by DillonJ** [comment]
Something like this, perhaps:

![image](https://user-images.githubusercontent.com/7080191/167438853-31583370-601f-4a72-8d0f-1b6f7f789059.png)

Edit - you could even omit the user constraint here actually

**Reply by LuckyLuke567** [reply]
Thank you, both the options you propose likely do the job. Unfortunately, I am not yet so confident with Spine to further comment these options, but I'll get to that.

I was also thinking about another option (in figure): two dummy technologies (no investment needed) taking as input, respectively, H2 and NG and producing a uniform energy output (that would be the blend). The fuel cell would then feed on the blend, which could have any composition. 

![image](https://user-images.githubusercontent.com/105210145/167567552-a97a00bd-8aab-4ea9-bae2-08146d7fc7ae.png)


Of course, besides other problems I probably overlooked, here it would be hard to define the efficiency of the fuel cell as a function of the blend.

**Reply by DillonJ** [reply]
Yes, that approach could work, and in fact, that is what I started out with... but in fact, I don't think the addional Fuel Cell unit and Uniform energy input node actually do anything. If the Dummy H2 and NG units are not investment decisions, then the blend is an operational decision and not an investment decisions as that is what I thought you were trying to achieve.

If you can express the efficiency as an affine linear function of the problem variables, then it is totally possible to represent the efficiency as a function of the blend. And in my experience, there is usually always a sufficiently acceptable linear formulation. If it's not linear, you will have problems in any modelling tool. But I suspect that by using the `fix_ratio_unit_flow_out_in` parameter to define the relationship between the input from NG and the flow to Uniform Energy Input and another `fix_ratio_unit_flow_out_in` parameter to define the relationship between the input from H2 and the flow to Uniform Energy Input you will effectively have an efficiency that is a function of the blend. If the relationship is more complex, that will be possible using user constraints. Perhaps we can assist you formulating the relationship you want?

I would love to hear more about why you are "not yet so confident with Spine" :-)

**Reply by LuckyLuke567** [reply]
Yep, now I see how it works. Indeed, the last solution you proposed renders the Uniform Energy Input node and the Fuel Cell unit superfluous.

My initial idea was to _"concentrate"_ the investment decision on the fuel cell (my question to the model is: is it convenient to invest in the fuel cells, and how much?), whereas the fuel costs would have been allocated to the dummy units. But again, your last solution makes this unnecessary, I believe.

The reason I am not yet so confident with Spine is that I started working on it just a couple of days ago. So, basically my fault, the documentation and case studies presented are actually very comprehensive. On this topic, sorry if I am posting on the community while I am a little unprepared, but I am in a position in which I have to select rapidly a tool, and this _fuel switching_ capability is crucial for my project.

**Reply by DillonJ** [reply]
No problem at all - we are happy to help - and if we can help further, please let us know. The idea in Spine is that using nodes, units and connections, the user can define their own technologies at a lower level, so the system is very flexible. It was designed from the ground up to model the new possibilities and interactions that arise in the context of the energy transition.
